### ISSUE TITLE
Hide @ModelAttribute parameters depending on a method

### ISSUE THREAD

**Issue by MaksimOrlov** [springfox/springfox#1356]
- Feature Request.

Good afternoon.

I'd like to have possibility to hide some parameters in @ModelAttribute annotated class at some endpoints.

For example:

I have two methods:

addOrder(@ModelAttribute Order order) 
updateOrder(@ModelAttribute Order order)

class Order has Id property.

In addOrder request id must be hidden, but in updateOrder id property must not be hidden.

I suggest to implement some annotation, like @HideParameter("addOrder",....)

So, at the class level it will look like that:

class Order
{
....
@HideParameter("addOrder")
private Long Id;
....
}

For now, there is only one way, is hide such a property at all requests, and then use @ApiInmplicitParam at request method level, but, unfortunately, it can lead to  mistakes.

Related question:

https://github.com/springfox/springfox/issues/896
https://github.com/springfox/springfox/issues/807

**Comment by dilipkrish** [comment]
How are you making sure that in the `updateOrder(@ModelAttribute Order order)` method the consumer is not sending you back an order id?

**Comment by MaksimOrlov** [comment]
If I correctly understand your question, there is a validation rule with 'group' parameter. 

``
@Null(groups=create.class)
@NotNull(groups=update.class)
private Long Id:
``
And at the controller level:
``
@Request mapping...
public updateOrder(@Validated(groups=update.class) Order order)
``

**Comment by dilipkrish** [comment]
👌 perfect! Why not use the bean validation groups to determine that instead of creating a new annotation?

**Comment by MaksimOrlov** [comment]
Yes, I'm completely agree with you, it's a good idea to use bean validation groups to determine visibility of a parameter.

If parameter has @Null(gropus={.....}) annotation for some group or  just @Null without groups (it means, that it's just calculated property or this parameter is used for the return response), it must not be present at the generated swagger file for related endpoint or at all. 

As I know, for now JSR-303 for request parameters is not supported yet, and is being under work. (https://github.com/springfox/springfox/pull/1254)
Correct me if I'm not right about that.

**Comment by MaksimOrlov** [comment]
Good evening, Dilip.

I'd like to suggest you a few ideas about that question:

The first, is to add _group validation_ support to PR #1371. So, each validation plugin (**NotNull**, **Size**, **MinMax**) will be applied when it is "switched on" by corresponding _group_ at _@ModelAttribute_ annotation level.

The second, is to add support for **@Null** annotation. It could be realised in the same way as all another. It will be related to _visibility_ property of the parameter builder class.

For group validation I would suggest that **_OperationParameterReader_** will read validation groups at the **@ModelAttribute** annotation level and then pass them in to **_ModelAttributeParameterExpander_**.

Then **_ModelAttributeParameterExpander_** will pass them in to **_ParameterExpansionContext_**. 
After that, validation groups can be read and applied by **_ExpandedParameterBuilderPlugin_**.


P.S. I can prepare all this changes and add to existing PR.

**Comment by MaksimOrlov** [comment]
I think, there is one little thing...
If we add supporting for @Null annotation with groups for Models (@RequestBody classes) then accordingly to #895 model parameter will be hidden at both request and response models if the same class was used.

May be we should find some way to split response and request for the same classes.

**Comment by dilipkrish** [comment]
@MaksimOrlov Let me 💭 about this and get back

**Comment by MaksimOrlov** [comment]
Good afternoon, Dilip.

I think I found some way, to resolve problem with connectivity between request and response for the same models.

For now, validation rules are beeing applyed regardless of request or response. But, inside Spring framework they are triggered only when request is processed.  (e.g. _@Min_ or _@Max_ annotations SpringFox can convert for the range for request parameter, but this limitation may be not the same after service logic will be processed and response model property may have some new range.)

**_ModelContext_** class has **boolean** property **_returnType_**, if I correctly understand, this property determines if model is used for response or not. 

Then in **_OptimizedModelPropertiesProvider_** this property can be passed to **_ModelPropertyContext_**.

After that, at the each **_ModelPropertyBuilderPlugin_**, that is corresponding to bean validation this property can be checked:

```
    @Override
    public void apply(ModelPropertyContext context) {
        
        if (context.isReturnType()) {
            return;
        }
        
        Optional<NotNull> notNull = extractAnnotation(context);
        context.getBuilder().required(notNull.isPresent());
    }
 ```

This code will prevent triggering validation rules for the response models.

**Comment by MaksimOrlov** [comment]
There is one more disappointing thing:

The default values of _@ApiModelProperty_ and _@ApiParam_ will override values of the _@Null_ and _@NotNull_ annotation and bring to nothing all benefits of the validation groups. 

I know that question had been discoursed before (#1231) than I opened this, but as I understand the problem is still present and swagger annotations impolite plugin order hasn't been implemented. Correct me, please, if I'm not right :)

The use case is:

```
pulic class Order()
{
......
@ApiModelProperty("Id of the user.")
@Digits(fraction = 0, integer = 1000)
@Min(value=1)
@NotNull(groups = ManagerCreate.class)
@Null(groups = {UserUpdate.class, UserCreate.class, ManagerUpdate.class})
private Integer User;
......
setters
....
getters
}
}
```

```
@PreAuthorize("hasRole('ROLE_USER')")
@RequestMapping(method=RequestMethod.POST, value="/addorder", params="uid")
public @ResponseBody Object addOrder(@ModelAttribute @Validated({Default.class, UserCreate.class}) NewOrderWithShifts newOrder)
{
....
....

}
```

```
NewOrderWithShifts extends Order;
```


Parameter **user** will be visible in the documentation because _@ApiModelProperty_ has default value _hidden=false_;

but if I add _hidden=true_ it will hide this parameter for all endpoints.

As I think there should be conjunction of the swagger and jsr annotation.

**Comment by dilipkrish** [comment]
@MaksimOrlov 

> ModelContext class has boolean property returnType, if I correctly understand, this property determines if model is used for response or not.

That is exactly right! that was the intension of that property. However [there is a catch](https://github.com/springfox/springfox/blob/8c44c62c0ba1d2b832b321aac6d7ec2ca6462a91/springfox-spring-web/src/main/java/springfox/documentation/spring/web/scanners/ApiModelReader.java#L82)... While its not relevant to ModelAttributes in itself, I think its the key to solving the problem long term i.e support `@JsonViews`, the concept of `Projections` in spring-data-rest etc. As a bonus the fact that we may have models with the same name in different packages.

I think couple of things need to happen to get it right
- Model names need to use the request/response type information
- Model names should use package names in addition to just the class names
- Request/Response objects should be collapsed to one only if they do not differ in properties, required (currently it just merges it into one always).

**Comment by MaksimOrlov** [comment]
@dilipkrish 

Sorry, I didn't pay attention for **Model merging**. My request model was extended from **_Order_** class and my response was just **_Order_**. So the Model names were different from each other and I didn't get in to that trap :)

Let me share some ideas about that:  

  &nbsp;
  
> - Model names need to use the request/response type information

&nbsp;

We can add a new property (for example "isUsedForResponse") to **Model** class and then pass returnType flag to it from ModelContext class. The functionality could be added to 

```
com.google.common.base.Optional<Model> DefaultModelProvider.modelFor(ModelContext modelContext)
```

Then model merging will be processed accordingly that flag.

&nbsp;
&nbsp;

> - Model names should use package names in addition to just the class names

&nbsp;

As I understand, simply putting _@ApiModel(**unique_model_name**)_ on model classes can solve [this](https://github.com/springfox/springfox/issues/182) issue :)

Or we can simply change  **_type.getSimpleName(_**) to **_type.getCanonicalName()_**  at the **_DefaultTypeNameProvider_**.

```
 String typeName = typeNameExtractor.typeName(ModelContext.fromParent(modelContext, propertiesHost));
    modelContext.getBuilder()
        .id(typeName)
```

The **_id_** parameter is used for the identifying models. Adding package name should solve problem.

Other option is to use package additions to names internally, but will swagger-ui  work correctly if it gets a few models with the same names?

And, I think that looks a little bit ugly at the swagger-ui

![swagger model names](https://cloud.githubusercontent.com/assets/7836239/18489616/e9f1ea6c-7a06-11e6-91ee-b3179784341e.png)

&nbsp;
&nbsp;

> - Request/Response objects should be collapsed to one only if they do not differ in properties, required (currently it just merges it into one always).

&nbsp;

```
private void mergeModelMap(Map<String, Model> target, Model source) 
```

should do this "magic" stuff and consider difference in properties and difference of the properties of the properties (E.g. required, range and etc.) 
But there is a new trap....
Names of this models will be remained the same, what will do with that swagger-ui?

&nbsp;

And,  I think, we should solve the priority problem. 
If the swagger annotation will be performed at the end, that will break group validation and @JsonView stuff, because swagger annotation will be applied at every method regardless of the conditions and bring back their  default values.

**Comment by MaksimOrlov** [comment]
The other way is to calculate **model**'s _hash_ depending on **_ModelContext.returnType_**, all _hashes_ of the **properties** and model **canonical name**. Then move this hash to model's **id**. Models's **name** could be changed with name_1 if the same was found as you mentioned at [this topic](https://github.com/springfox/springfox/issues/182#issuecomment-167989428).

**Comment by MaksimOrlov** [comment]
@dilipkrish should I prepare some initial code changes for the model hashes and model names?

**Comment by dilipkrish** [comment]
Sure that would be awesome, that would be a great help! Also as a background . @mbazos tried to solve this problem, here is the issue #182 with a large discussion thread

To restate the problem 
- we need to solve this problem for 2 cases, one for model attributes, and one for regular models that are either request body or response body models.
- we need to use this solution for supporting @JsonView
- We potentially will need to be able to use this for supporting projections in spring-data-rest
- Same class name different package

Would be awesome if you would create short writeup how you plan to approach it and

**Comment by dilipkrish** [comment]
Not that you need to solve all those problems 😉, but your solution should allow for them to be solved with minimal effort.

**Comment by MaksimOrlov** [comment]
Good afternoon, @dilipkrish !

I've made some initial code changes.

The short write-up is:

1. Add new option **ModelComparingEnabled** to **Docket** ,  **DocumentationContext** and **DocumentationContextBuilder**.

2. According to this option models will be merged or their type and properties will be compared.

3. If after comparing we determine that the model is different from previous, the **ModelNameGenerator** will add digit to the end of the name.

4. Then model will be added to others.

**Comment by MaksimOrlov** [comment]
Just let me know if I'm on a right way :)

**Comment by MaksimOrlov** [comment]
Good evening, Dilip.

May be this issue should not be closed, just transferred to 2.7.0 version?

**Comment by dilipkrish** [comment]
Thanks for catching that... was bulk updating :)

**Comment by MaksimOrlov** [comment]
@dilipkrish, would you mind answering for a few questions

Do we really need _model merging_? May be we should completely substitute it by _model comparing_?

What do you think, is it a appropriate way to realize **_.equals_** method for the model class and then use it at the **ApiModelReader** class ?

P.S. This issue is still closed... Could you, please, reopen it?

**Comment by dilipkrish** [comment]
Thot I reopened... sorry.

We don't need model merging if we solve the problem that request and response are different. Otherwise what will end up happening is that, either one of those models will be incorrect, based on which came first

for e.g.
1. We read model as `Model get(id);` first
1. We read model as `void put(Model);` second

Now because this model reading is cached, we'll never get the model right for one or the other, depending on the order.

If we figure out that problem, we don't need model merging at all. makes sense?

**Comment by MaksimOrlov** [comment]
Yes, @dilipkrish, it does make sense.

I've done some changes for this issue (https://github.com/MaksimOrlov/springfox/tree/model-comparing), and now I'm working on unit tests and code cleanings. 

The main idea for that changes is: to add link between **Parameters** and **Models**. Without this connection could be the situation when model's _name_ was changed (i.e. _name_ ->_name_1_), but **ModelReference** (in **Parameter** or **ModelProperty**)  is still pointing to old model's _name_, that will lead to nonconforming documentation.

1. At the **OperationModelContextsBuilder** added **Map<ResolvedMethodParameter, ModelContext>** to keep connection between **Parameter** and **ModelContexts**.

2. Adjusted method _updateModelRef_ in **ModelProperty** to keep **ModelReferenceProvider** with the same **ModelContext** as **Model** was built from.

3. Added algorithm to compare models.

4. Model's index will be adjusted with **ModelContext** and **ModelBuilder** will save index.

5. **ModelReferenceProvider** will return reference depends on _model.getIndex()_

**Comment by dilipkrish** [comment]
@MaksimOrlov sorry have been a little busy lately. sounds complicated ☝looking fwd to it! Thanks for working on it, much appreciated.

**Comment by MaksimOrlov** [comment]
Good evening, @dilipkrish
I'm ready with a PR #1671 :)

**Comment by ltylty** [comment]
is this problem solved in 2.7.0 ? how to config?

**Comment by MaksimOrlov** [comment]
@ltylty, unfortunately, it is still under work...
I will finish with this feature in a next few months, I hope.

**Comment by MaksimOrlov** [comment]
We have in a PR (#1671) just raw model comparing (that will be reimplemented in some part), without bean validation groups.

**Comment by JcacMing** [comment]
@MaksimOrlov 
I wish you a success

**Comment by MaksimOrlov** [comment]
@JcacMing 
Thanks a lot!

**Comment by dilipkrish** [comment]
Related to #2056

**Comment by prabhushrikant** [comment]
Is the group validation actually working? I am using springfox 2.9.2 and seems to be not working.

My controller method looks like this  

`@ApiOperation(value = "Create new group.", notes = "Create a group in the org identified by orgId." +
      "If orgId is null, then the caller's agency is used.")
  @ApiResponses(value = {
      @ApiResponse(code = 201, message = "Newly created group.", response = Group.class),
      @ApiResponse(code = 400, message = MESSAGE_400),
      @ApiResponse(code = 403, message = MESSAGE_403),
      @ApiResponse(code = 406, message = MESSAGE_406),
      @ApiResponse(code = 500, message = MESSAGE_500)
  })
  @ResponseStatus(HttpStatus.CREATED)
  @RequestMapping(value = "/create-user", method = RequestMethod.POST)
  public ResponseEntity<User> createUser(
      @ApiParam(value = "Payload for user create request.", required = true)
      final @Validated @RequestBody User userToBeCreated)
      throws InvalidAttributeException, NotAuthorizedException, AmspeException,
          InvalidOperationException {
//... logic for creating new user 
}
`
My user object looks as follows
`
@Data
@Builder
@ToString
@NoArgsConstructor
public class User {
  public static final int USER_DESCRIPTION_MAXLEN = 200;
  public static final int USER_NAME_MAXLEN = 64;

  @NotNull(groups = {ValidationGroups.UpdateChecks.class, ValidationGroups.DeleteChecks.class},
  message = "User id can't be null.")
  @NotBlank(groups = {ValidationGroups.UpdateChecks.class, ValidationGroups.DeleteChecks.class},
  message = "User id can't be empty.")
  private UUID id;

  @NotNull(message = "User name can't be null.")
  @NotBlank(message = "User name can't be empty.")
  @Size(max = USER_NAME_MAXLEN, message = "Not a valid user name.")
  private String userName;

  private UUID owner;

  @Size(max = USER_DESCRIPTION_MAXLEN, message = "Not a valid user description.")
  private String description;

  @NotNull
  @NotBlank
  private UUID agency;

  /**
   * All args constructor
   */
  public User(final UUID id, final String groupName, final UUID owner, final String description,
                final UUID agency) {
    this.id = id;
    this.groupName = groupName;
    this.owner = owner;
    this.description = description;
    this.agency = agency;
  }`

My swagger documentation still shows user model as follows , where id is shown as mandatory filed for create user request , where as it's mandatory for only updateChecks and deleteChecks groups.
![image](https://user-images.githubusercontent.com/6848200/53922864-e79ad980-402a-11e9-906e-7a78db14a0ff.png)

**Comment by MaksimOrlov** [comment]
@prabhushrikant, validation groups feature is not supported by current version. It is planed from 3.x releases.

**Comment by MaksimOrlov** [comment]
PR for [that](https://github.com/springfox/springfox/pull/2056) is almost done.

**Comment by MaksimOrlov** [comment]
PR has been merged to master and available in 2.9.2-snapshot version.

**Comment by zhaomoran** [comment]
@MaksimOrlov I'm trying to use 3.0.0-SNAPSHOT, The problem reported in @prabhushrikant  still exists.

**Comment by ltylty** [comment]
is this fixed in 3.0 version? how to config?

**Comment by ltylty** [comment]
knife4j solve this problem
https://doc.xiaominfo.com/knife4j/documentation/filterRequestParameter.html
